src/Storages/ObjectStorageQueue/ObjectStorageQueueSource.cpp

```diff
@@ -841,42 +845,72 @@ Chunk ObjectStorageQueueSource::generateImpl()
 
         LOG_TEST(log, "Processing file: {}", path);
 
+        Chunk chunk;
+        bool result = false;
+
         try
         {
-            auto timer = DB::CurrentThread::getProfileEvents().timer(ProfileEvents::ObjectStorageQueuePullMicroseconds);
-
-            Chunk chunk;
-            if (reader->pull(chunk))
+            if (file_status->processed_rows == 0 && fault_injection_probability > 0.)
             {
-                LOG_TEST(log, "Read {} rows from file: {}", chunk.getNumRows(), path);
-
-                file_status->processed_rows += chunk.getNumRows();
-                progress->processed_rows += chunk.getNumRows();
-                progress->processed_bytes += chunk.bytes();
-
-                ProfileEvents::increment(ProfileEvents::ObjectStorageQueueReadRows, chunk.getNumRows());
-                ProfileEvents::increment(ProfileEvents::ObjectStorageQueueReadBytes, chunk.bytes());
+                std::bernoulli_distribution fault(fault_injection_probability);
+                if (fault(thread_local_rng))
+                    throw Exception(ErrorCodes::UNKNOWN_EXCEPTION, "Failed to pull chunk from file");
+            }
 
-                VirtualColumnUtils::addRequestedFileLikeStorageVirtualsToChunk(
-                    chunk, read_from_format_info.requested_virtual_columns,
-                    {
-                        .path = path,
-                        .size = reader.getObjectInfo()->metadata->size_bytes
-                    }, getContext());
+            auto timer = DB::CurrentThread::getProfileEvents().timer(ProfileEvents::ObjectStorageQueuePullMicroseconds);
 
-                return chunk;
-            }
+            result = reader->pull(chunk);
         }
         catch (...)
         {
             const auto message = getCurrentExceptionMessage(true);
-            LOG_ERROR(log, "Got an error while pulling chunk. Will set file {} as failed. Error: {} ", path, message);
+            LOG_ERROR(
+                log, "Got an error while pulling chunk. Will set file {} as failed "
+                "(processed rows: {}). Error: {} ",
+                path, file_status->processed_rows, message);
 
             processed_files.back().state = FileState::ErrorOnRead;
             processed_files.back().exception_during_read = message;
 
-            /// Stop processing and commit what is already processed.
-            return {};
+            if (file_status->processed_rows > 0)
```

Comment:
This used to be here before, but got lost in the multiple code rewrites in 25.1.
For 24.12 branch:
https://github.com/ClickHouse/ClickHouse/blob/6ab59f4355de0290053afb5913f7e3763366f469/src/Storages/ObjectStorageQueue/ObjectStorageQueueSource.cpp#L592